### PR TITLE
Pulling function registration to a separate registry.

### PR DESCRIPTION
**Pull request by jeeyoungk**
* registry interface is pulled out.
* Renaming some of the functions (transformX -> X)
* Function registry is no longer global during the execution of the uery (but it's not configurable yet)

@Nathan-Fenner @achow

### CODE REVIEW

File: function/expression.go [hunk 1]
```diff
@@ -23,6 +24,35 @@ type EvaluationContext struct {
...
+	Compute     func(EvaluationContext, []Expression, []string) (Value, error)
+}
+
+// Evaluate the given metric function.
+func (f MetricFunction) Evaluate(context EvaluationContext,
+	arguments []Expression, groupBy []string) (Value, error) {
+	// preprocessing
+	length := len(arguments)
+	if length < f.MinArgument || (f.MaxArgument != -1 && f.MaxArgument < length) {
+		return nil, ArgumentLengthError{f.Name, f.MinArgument, f.MaxArgument, length}
```
Comment (achow) [inline function/expression.go]: Should supply field names when building the error struct.

File: function/expression.go [hunk 2]
```diff
@@ -23,6 +24,35 @@ type EvaluationContext struct {
...
+}
+
+type Registry interface {
+	GetFunction(string) (MetricFunction, bool)
+}
+
+// MetricFunction defines a common logic to dispatch a function in MQE.
+type MetricFunction struct {
+	Name        string
+	MinArgument int
```
Comment (achow) [inline function/expression.go]: I'd prefer `MinArguments` and `MaxArguments`

File: function/registry/registry.go [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+	MustRegister(NewFilter("filter.highest_min", aggregate.Min, false))
+	MustRegister(NewFilter("filter.lowest_min", aggregate.Min, true))
+}
+
+// StandardRegistry of a functions available in MQE.
+type StandardRegistry struct {
+	mapping map[string]function.MetricFunction
+}
+
+var defaultRegistry = StandardRegistry{make(map[string]function.MetricFunction)}
```
Comment (achow) [inline function/registry/registry.go]: Should supply field name when creating the struct.

File: function/expression.go [hunk 4]
```diff
@@ -23,6 +24,35 @@ type EvaluationContext struct {
...
+	MaxArgument int
+	Groups      bool // Whether the function allows a 'group by' clause.
```
Comment (achow) [inline function/expression.go]: Should name this `AllowsGroups`

File: function/expression.go (hunk 1 above)
Comment (jeeyoungk) [inline function/expression.go]: I don't think we have argument names currently.

This PR is mostly move, so I'll add it later.

File: function/expression.go (hunk 2 above)
Comment (jeeyoungk) [inline function/expression.go]: done.

File: function/registry/registry.go (hunk 3 above)
Comment (jeeyoungk) [inline function/registry/registry.go]: done

File: function/expression.go (hunk 4 above)
Comment (jeeyoungk) [inline function/expression.go]: done

**Comment by achow** [comment]
LGTM

**Comment by Nathan-Fenner** [comment]
LGTM

**Pull request merged**